pkg/security/ptracer/ptracer.go

```diff
@@ -239,17 +240,110 @@ func (t *Tracer) ReadArgStringArray(pid int, regs syscall.PtraceRegs, arg int) (
 	return result, nil
 }
 
-// Trace traces a process
-func (t *Tracer) Trace(cb func(cbType CallbackType, nr int, pid int, ppid int, regs syscall.PtraceRegs, waitStatus *syscall.WaitStatus)) error {
+func isExited(waitStatus syscall.WaitStatus) bool {
+	return waitStatus.Exited() || waitStatus.CoreDump() || waitStatus.Signaled()
+}
+
+func (t *Tracer) trace(cb func(cbType CallbackType, nr int, pid int, ppid int, regs syscall.PtraceRegs, waitStatus *syscall.WaitStatus)) error {
+	var waitStatus syscall.WaitStatus
+
+	if err := syscall.PtraceSyscall(t.PID, 0); err != nil {
+		return err
+	}
+
+	var (
+		tracker = NewSyscallStateTracker()
+		regs    syscall.PtraceRegs
+	)
+
+	for {
+		pid, err := syscall.Wait4(-1, &waitStatus, 0, nil)
+		if err != nil {
+			t.opts.Logger.Debugf("unable to wait for pid %d: %v", pid, err)
+			break
+		}
+
+		if isExited(waitStatus) {
+			tracker.Exit(pid)
+
+			if pid == t.PID {
+				break
+			}
+			cb(CallbackExitType, ExitNr, pid, 0, regs, &waitStatus)
+			continue
+		}
+
+		if waitStatus.Stopped() {
+			if signal := waitStatus.StopSignal(); signal != syscall.SIGTRAP {
+				if signal == syscall.SIGSTOP {
+					signal = syscall.Signal(0)
+				}
+				if err := syscall.PtraceSyscall(pid, int(signal)); err == nil {
+					continue
+				}
+			}
+
+			if err := syscall.PtraceGetRegs(pid, &regs); err != nil {
+				t.opts.Logger.Debugf("unable to get registers for pid %d: %v", pid, err)
+
+				// it got probably killed
+				cb(CallbackExitType, ExitNr, pid, 0, regs, &waitStatus)
+
+				if pid == t.PID {
+					break
+				}
+				continue
+			}
+
+			nr := GetSyscallNr(regs)
+
+			switch waitStatus.TrapCause() {
+			case syscall.PTRACE_EVENT_CLONE, syscall.PTRACE_EVENT_FORK, syscall.PTRACE_EVENT_VFORK:
+				if npid, err := syscall.PtraceGetEventMsg(pid); err == nil {
+					cb(CallbackPostType, nr, int(npid), pid, regs, nil)
+				}
+			case syscall.PTRACE_EVENT_EXEC:
+				cb(CallbackPostType, ExecveNr, pid, 0, regs, nil)
+
+				if state := tracker.PeekState(pid); state != nil {
+					state.Exec = true
+				}
```

Comment:
I'm not sure it's possible, but if the FIRST syscall a prog make is an exec (like, directly after the exec launching it), the state won't be initialized yet here (by NextStop()) and PeekState will return nil

Reply:
here that is post of an exec, the entry is line 316, so we should have already a NextStop for this pid. I'm not sure that the functional tests will pass if it wasn't the case. Did I miss something regarding your comment ?

Reply:
You're right, I misread this part :)